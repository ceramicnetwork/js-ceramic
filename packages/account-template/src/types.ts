import type Ceramic from '@ceramicnetwork/ceramic-core'
import type CeramicClient from '@ceramicnetwork/ceramic-http-client'
import type CoreDocument from '@ceramicnetwork/ceramic-core/lib/document'
import type ClientDocument from '@ceramicnetwork/ceramic-http-client/lib/document'

export type CeramicApi = Ceramic | CeramicClient;

// Because Ceramic and CeramicClient use different Document definition:
export type CeramicDocument = CoreDocument | ClientDocument;
