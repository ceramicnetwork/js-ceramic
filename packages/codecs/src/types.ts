// Modified version of ts-essentials `Opaque`.
// Added WithBrand interface to overcome TS issue #37888.

type StringLiteral<T> = T extends string ? (string extends T ? never : T) : never

declare const __BRAND__: unique symbol

export interface WithBrand<Token extends string> {
  [__BRAND__]: Token
}

export type Opaque<Type, Brand extends string> = Brand extends StringLiteral<Brand>
  ? Type & WithBrand<Brand>
  : never
