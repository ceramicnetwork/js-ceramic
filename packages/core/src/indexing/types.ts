import type { StreamID } from '@ceramicnetwork/streamid'

export interface IndexStreamArgs {
  streamID: StreamID
  model: StreamID
  controller: string
  lastAnchor: Date | null
}

export enum Ordering {
  CHRONOLOGICAL = 'chronological', // last_anchored_at DESC NULLS FIRST, created_at DESC
  INSERTION = 'insertion', // created_at DESC = when an entry was added to the index
}

export interface BaseQuery {
  model: StreamID | string
  account?: string
  order?: Ordering // default: CHRONOLOGICAL_DESC
}

export interface ForwardPagination {
  first: number
  after?: string
}

export interface BackwardPagination {
  last: number
  before?: string
}

export type Pagination = ForwardPagination | BackwardPagination

export interface Page<T> {
  entries: Array<T>
  pageInfo: PageInfo
}

export type PageInfo = {
  hasNextPage: boolean
  hasPreviousPage: boolean
  startCursor: string
  endCursor: string
}

// Per database
export interface DatabaseIndexAPI {
  /**
   * This method inserts the stream if it is not present in the index, or updates
   * the 'content' if the stream already exists in the index.
   * @param args
   */
  indexStream(args: IndexStreamArgs): Promise<void>

  /**
   * Query the index
   */
  page(query: BaseQuery & Pagination): Promise<Page<StreamID>>

  /**
   * Initialize connection to a database.
   */
  init(): Promise<void>
}
