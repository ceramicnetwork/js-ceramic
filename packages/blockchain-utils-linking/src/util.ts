export interface LinkProof {
  version: number;
  message: string;
  signature: string;
  account: string;
  did?: string;
  timestamp?: number;
  address?: string;
  type?: string;
  chainId?: number;
}

export interface RpcMessage {
  jsonrpc: string;
  id: number;
  method: string;
  params: any;
}

export interface ConsentMessage {
  message: string;
  timestamp?: number;
}

export function getConsentMessage(
  did: string,
  addTimestamp = true
): ConsentMessage {
  const res: any = {
    message: "Create a new account link to your identity." + "\n\n" + did,
  };
  if (addTimestamp) {
    res.timestamp = Math.floor(Date.now() / 1000);
    res.message += " \n" + "Timestamp: " + res.timestamp;
  }
  return res;
}

export function encodeRpcMessage(method: string, params?: any): any {
  return {
    jsonrpc: "2.0",
    id: 1,
    method,
    params,
  };
}
