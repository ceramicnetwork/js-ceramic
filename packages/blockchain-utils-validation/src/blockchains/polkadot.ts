import { BlockchainHandler } from "../blockchain-handler";
import { AccountID } from "caip";
import { signatureVerify } from "@polkadot/util-crypto";
import { LinkProof } from "@ceramicnetwork/blockchain-utils-linking";

const namespace = "polkadot";

async function validateLink(proof: LinkProof): Promise<LinkProof | null> {
  const address = new AccountID(proof.account).address;
  const res = await signatureVerify(proof.message, proof.signature, address);
  return res.isValid ? proof : null;
}

const Handler: BlockchainHandler = {
  namespace,
  validateLink,
};

export default Handler;
