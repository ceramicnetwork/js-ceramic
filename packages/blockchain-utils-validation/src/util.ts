import { AccountId } from 'caip'

export function normalizeAccountId(accountId: string | AccountId): AccountId {
  // Handle legacy CAIP-10 representation
  if (typeof accountId === 'string' && accountId.includes('@')) {
    const [address, chainId] = accountId.split('@')
    if (!address || !chainId) {
      throw new Error(`Invalid accountId provided`)
    }

    return new AccountId({ address, chainId })
  }
  return new AccountId(accountId)
}
