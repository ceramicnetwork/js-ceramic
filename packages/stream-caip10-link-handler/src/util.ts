// NOTE: This differs from `normalizeAccountId` in `stream-caip10-link`
// This converts from CAIP-10 repr to the legacy repr
// The former converts from legacy to the current repr
export function normalizeAccountId(accountId: string): string {
  if (!accountId.includes('@')) {
    const _accountSplit = accountId.split(':')
    const address = _accountSplit.pop()
    const chainId = _accountSplit.join(':')
    return `${address}@${chainId}`
  }
  return accountId
}
