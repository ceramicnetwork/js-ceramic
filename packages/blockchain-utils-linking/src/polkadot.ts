import { AccountID } from "caip";
import { AuthProvider } from "./auth-provider";
import { getConsentMessage, LinkProof } from "./util";
import * as uint8arrays from "uint8arrays";

const stringHex = (str: string): string =>
  `0x${uint8arrays.toString(uint8arrays.fromString(str), "base16")}`;

const CHAIN_ID = "polkadot:b0a8d493285c2df73290dfb7e61f870f";

export class PolkadotAuthProvider implements AuthProvider {
  readonly isAuthProvider = true;

  constructor(
    private readonly provider: any,
    private readonly address: string
  ) {}

  async authenticate(message: string): Promise<string> {
    throw new Error(`Not Implemented: PolkadotAuthProvider#authenticate, ${message}`);
  }

  async createLink(did: string): Promise<LinkProof> {
    const { message, timestamp } = getConsentMessage(did);
    const linkMessageHex = stringHex(message);
    const account = await this.accountId();
    const res = await this.provider.signRaw({
      address: this.address,
      data: linkMessageHex,
      type: "bytes",
    });
    return {
      version: 2,
      type: "eoa",
      message: linkMessageHex,
      signature: res.signature,
      account: account.toString(),
      timestamp: timestamp,
    };
  }

  async accountId(): Promise<AccountID> {
    return new AccountID({ address: this.address, chainId: CHAIN_ID });
  }

  withAddress(address: string): AuthProvider {
    return new PolkadotAuthProvider(this.provider, address);
  }
}
