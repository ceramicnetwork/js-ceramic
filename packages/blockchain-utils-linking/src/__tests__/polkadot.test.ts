import { Signer, SignerResult } from "@polkadot/api/types";
import { KeyringPair } from "@polkadot/keyring/types";
import { SignerPayloadRaw } from "@polkadot/types/types";
import { TypeRegistry } from "@polkadot/types/create";
import { createTestKeyring } from "@polkadot/keyring/testing";
import { assert, hexToU8a, u8aToHex } from "@polkadot/util";
import { PolkadotAuthProvider } from "../polkadot";
import { cryptoWaitReady } from "@polkadot/util-crypto";

const did = "did:3:bafysdfwefwe";
const seed = hexToU8a(
  "0xabf8e00000000000000000000000000000000000000000000000000000000000"
);

// primary and supported by polkadot extension
const keyringSr25519 = createTestKeyring({ type: "sr25519" });
const keyringEd25519 = createTestKeyring({ type: "ed25519" });
const keyringSecp256k = createTestKeyring({ type: "ecdsa" });

class SingleAccountSigner implements Signer {
  readonly #keyringPair: KeyringPair;

  constructor(registry: TypeRegistry, keyringPair: KeyringPair) {
    this.#keyringPair = keyringPair;
  }

  public async signRaw({
    address,
    data,
  }: SignerPayloadRaw): Promise<SignerResult> {
    assert(
      address === this.#keyringPair.address,
      "Signer does not have the keyringPair"
    );

    return new Promise((resolve): void => {
      const signature = u8aToHex(this.#keyringPair.sign(hexToU8a(data)));
      resolve({ id: 1, signature });
    });
  }
}

const registry = new TypeRegistry();
let keyPairSr25519: KeyringPair,
  keyPairEd25519: KeyringPair,
  keyPairSecp256k: KeyringPair;

beforeAll(async () => {
  await cryptoWaitReady();
  keyPairSr25519 = keyringSr25519.addFromSeed(seed);
  keyPairEd25519 = keyringEd25519.addFromSeed(seed);
  keyPairSecp256k = keyringSecp256k.addFromSeed(seed);
  global.Date.now = jest.fn().mockImplementation(() => 666000);
});

afterAll(() => {
  jest.clearAllMocks();
});

test("accountId", async () => {
  const provider = new SingleAccountSigner(registry, keyPairSr25519);
  const authProvider = new PolkadotAuthProvider(
    provider,
    keyPairSr25519.address
  );
  await expect(authProvider.accountId()).resolves.toMatchSnapshot();
});

describe("createLink", () => {
  test("create proof with sr25519", async () => {
    const provider = new SingleAccountSigner(registry, keyPairSr25519);
    const authProvider = new PolkadotAuthProvider(
      provider,
      keyPairSr25519.address
    );
    const proof = await authProvider.createLink(did);
    expect(proof.account).toMatchSnapshot();
    expect(proof.message).toMatchSnapshot();
    expect(proof.type).toMatchSnapshot();
  });
  test("create proof with ed25519", async () => {
    const provider = new SingleAccountSigner(registry, keyPairEd25519);
    const authProvider = new PolkadotAuthProvider(
      provider,
      keyPairEd25519.address
    );
    const proof = await authProvider.createLink(did);
    expect(proof).toMatchSnapshot();
  });
  test("create proof with secp256k", async () => {
    const provider = new SingleAccountSigner(registry, keyPairSecp256k);
    const authProvider = new PolkadotAuthProvider(
      provider,
      keyPairSecp256k.address
    );
    const proof = await authProvider.createLink(did);
    expect(proof).toMatchSnapshot();
  });
  test("address mismatch", async () => {
    const provider = new SingleAccountSigner(registry, keyPairEd25519);
    const authProvider = new PolkadotAuthProvider(
      provider,
      keyPairSecp256k.address
    );
    await expect(authProvider.createLink(did)).rejects.toThrow();
  });
});

test("authenticate", async () => {
  const provider = new SingleAccountSigner(registry, keyPairEd25519);
  const authProvider = new PolkadotAuthProvider(
    provider,
    keyPairSecp256k.address
  );
  await expect(authProvider.authenticate("msg")).rejects.toThrow(
    "Not Implemented"
  );
});
