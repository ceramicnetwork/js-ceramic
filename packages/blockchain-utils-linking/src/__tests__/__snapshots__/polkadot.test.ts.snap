// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accountId 1`] = `
AccountID {
  "address": "5CXU7u72Gz9Jqo6FdAwzSsxX5JhdRwMj2UZ4xrAEoo3AyySx",
  "chainId": ChainID {
    "namespace": "polkadot",
    "reference": "b0a8d493285c2df73290dfb7e61f870f",
  },
}
`;

exports[`createLink create proof with ed25519 1`] = `
Object {
  "account": "5EEvMbuaJQxxPorGtzY2qUYqM3akRrnD36bZD4PomfpLdrtx@polkadot:b0a8d493285c2df73290dfb7e61f870f",
  "message": "0x4372656174652061206e6577206163636f756e74206c696e6b20746f20796f7572206964656e746974792e0a0a6469643a333a626166797364667765667765200a54696d657374616d703a20363636",
  "signature": "0xe7c359daca1d1c2bc82116654ef8277b989624e833966969c556fc82b12875693cb850c64bbb3f145a7db46d9b1bcf04b606960be89d0daac07945791557db0a",
  "timestamp": 666,
  "type": "eoa",
  "version": 2,
}
`;

exports[`createLink create proof with secp256k 1`] = `
Object {
  "account": "5GuAnwx9uKSQxKZB8moxoaX4ceehoCYu4XYMjJkKtm2QDgqC@polkadot:b0a8d493285c2df73290dfb7e61f870f",
  "message": "0x4372656174652061206e6577206163636f756e74206c696e6b20746f20796f7572206964656e746974792e0a0a6469643a333a626166797364667765667765200a54696d657374616d703a20363636",
  "signature": "0x74685cde417eac26ee6d35667772abd9f1243b7c04973848f7b0332a8eca52af37c5d6ece58d73c72b0accbba65a0361fa4d3c2812b761200c87cfdd52c2f15000",
  "timestamp": 666,
  "type": "eoa",
  "version": 2,
}
`;

exports[`createLink create proof with sr25519 1`] = `"5CXU7u72Gz9Jqo6FdAwzSsxX5JhdRwMj2UZ4xrAEoo3AyySx@polkadot:b0a8d493285c2df73290dfb7e61f870f"`;

exports[`createLink create proof with sr25519 2`] = `"0x4372656174652061206e6577206163636f756e74206c696e6b20746f20796f7572206964656e746974792e0a0a6469643a333a626166797364667765667765200a54696d657374616d703a20363636"`;

exports[`createLink create proof with sr25519 3`] = `"eoa"`;
